internal/controller/postgrescluster/instance.go

```diff
@@ -1341,3 +1380,382 @@ func (r *Reconciler) reconcileInstanceCertificates(
 
 	return instanceCerts, err
 }
+
+// reconcileUpgradeJob creates the Postgres major upgrade Job based on
+// observations made about the cluster's status. Normal PostgresCluster instance
+// reconciliation is halted while the upgrade takes place. One portion of the
+// upgrade preparation process, deleting any existing Endpoints, is handled
+// directly in this method in order to allow for cluster reinitialization. As
+// the upgrade process proceeds, relevant status conditions are set to track the
+// progress being made. Once complete, normal cluster reconciliation is allowed
+// to continue.
+func (r *Reconciler) reconcileUpgradeJob(
+	ctx context.Context, cluster *v1beta1.PostgresCluster,
+	observed *observedInstances, instanceSets []v1beta1.PostgresInstanceSetSpec,
+	sa string, clusterCerts, clientCerts *corev1.SecretProjection,
+	clusterVolumes []corev1.PersistentVolumeClaim,
+) (bool, error) {
+
+	// TODO(tjmoore4): Checking the PG versions below, but in the future, version
+	// validation should be added as a webhook.
+	if cluster.Spec.Upgrade != nil && *cluster.Spec.Upgrade.Enabled {
+		if cluster.Spec.PostgresVersion > cluster.Spec.Upgrade.FromPostgresVersion {
+			// before creating the upgrade job, observe all resources currently
+			// relevant to reconciling data sources, and update status accordingly
+			endpoints, upgradeJob, err := r.observeUpgradeEnv(ctx, cluster)
+			if err != nil {
+				return false, errors.WithStack(err)
+			}
+
+			if upgradeJob == nil {
+				err = r.prepareForUpgrade(ctx, cluster, observed, endpoints, upgradeJob)
+				if err != nil {
+					return false, errors.WithStack(err)
+				}
+
+				// Return early until there are no observed instances, i.e. all
+				// postgres Pods are stopped or if there is an error.
+				observedInstances, err := r.observeInstances(ctx, cluster)
+				if err != nil {
+					return false, err
+				}
+				if len(observedInstances.byName) != 0 {
+					return true, nil
+				}
+
+				// After all observed instances are deleted, delete endpoints to
+				// remove the old DCS information.
+				if len(endpoints) > 0 {
+					for i := range endpoints {
+						if err := r.Client.Delete(ctx, &endpoints[i]); client.IgnoreNotFound(err) != nil {
```

Comment:
:pick: Feels like endpoint deletion might be better aligned under `prepareForUpgrade()` rather than this function.